src/code/sys_math3d.c

```diff
@@ -2,212 +2,376 @@
+                               f32 planeBC, f32 planeBDist, InfiniteLine* intersect);
+s32 Math3D_CirSquareVsTriSquare(f32 x0, f32 y0, f32 x1, f32 y1, f32 x2, f32 y2, f32 centerX, f32 centerY, f32 radius);
+s32 Math3D_SphCubeVsTriCube(Vec3f* v0, Vec3f* v1, Vec3f* v2, Vec3f* center, f32 radius);
+
+/**
+ * Creates an infinite line along the intersection of the plane defined from `planeAA`x + `planeAB`y + `planeAB`z +
+ * `planeADist` = 0 and `planeBA`x + `planeBB`y + `planeBC`z + `planeBDist` = 0, and finds the closest point on that
+ * intersection to the line segment `linePointA and linePointB`, outputs the intersection to `closestPoint`
+ */
+s32 Math3D_PlaneVsLineSegClosestPoint(f32 planeAA, f32 planeAB, f32 planeAC, f32 planeADist, f32 planeBA, f32 planeBB,
+                                      f32 planeBC, f32 planeBDist, Vec3f* linePointA, Vec3f* linePointB,
+                                      Vec3f* closestPoint) {
+    static InfiniteLine planeIntersectLine;
+    static Linef planeIntersectSeg;
 
-    Vec3f sp34;
+    Vec3f sp34; // unused
 
-    if (func_800CAD08(arg0, arg1, arg2, arg3, arg4, arg5, arg6, arg7, &D_8016A5A0) == 0) {
-        return 0;
+    if (!Math3D_PlaneVsPlaneNewLine(planeAA, planeAB, planeAC, planeADist, planeBA, planeBB, planeBC, planeBDist,
+                                    &planeIntersectLine)) {
+        // The planes are parallel
+        return false;
     }
-    Math_Vec3f_Copy(&D_8016A5B8.a, &D_8016A5A0.a);
 
-    D_8016A5B8.b.x = (D_8016A5A0.b.x * 100.0f) + D_8016A5A0.a.x;
-    D_8016A5B8.b.y = (D_8016A5A0.b.y * 100.0f) + D_8016A5A0.a.y;
-    D_8016A5B8.b.z = (D_8016A5A0.b.z * 100.0f) + D_8016A5A0.a.z;
+    // create a line segment on the plane.
+    Math_Vec3f_Copy(&planeIntersectSeg.a, &planeIntersectLine.point);
+    planeIntersectSeg.b.x = (planeIntersectLine.dir.x * 100.0f) + planeIntersectLine.point.x;
+    planeIntersectSeg.b.y = (planeIntersectLine.dir.y * 100.0f) + planeIntersectLine.point.y;
+    planeIntersectSeg.b.z = (planeIntersectLine.dir.z * 100.0f) + planeIntersectLine.point.z;
 
-    if (!func_800CA8E8(&D_8016A5B8.a, &D_8016A5B8.b, arg8, arg9, argA, &sp34)) {
-        return 0;
+    // closestPoint is a point on planeIntersect, sp34 is a point on linePointA, linePointB
+    if (!Math3D_LineSegMakePerpLineSeg(&planeIntersectSeg.a, &planeIntersectSeg.b, linePointA, linePointB, closestPoint,
+                                       &sp34)) {
+        return false;
     }
-    return 1;
+    return true;
 }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/sys_math3d/func_800CA8E8.s")
+#ifdef NON_MATCHING
+/**
+ * Creates a line segment which is perpendicular to the line segments `lineAPointA`->`lineAPointB` and
+ * `lineBPointA`->`lineBPointB`
+ *
+ * NON_MATCHING:
+ * Lots of regalloc, but is functionally equivilent, some reordering.
+ */
+s32 Math3D_LineSegMakePerpLineSeg(Vec3f* lineAPointA, Vec3f* lineAPointB, Vec3f* lineBPointA, Vec3f* lineBPointB,
```

Comment:
In this kind of situation I suggest splitting comments like this (and in general the function documentation should be **above** the ifdef since it's always relevant)
```suggestion
/**
 * Creates a line segment which is perpendicular to the line segments `lineAPointA`->`lineAPointB` and
 * `lineBPointA`->`lineBPointB`
 */
#ifdef NON_MATCHING
// Lots of regalloc, but is functionally equivilent, some reordering.
s32 Math3D_LineSegMakePerpLineSeg(Vec3f* lineAPointA, Vec3f* lineAPointB, Vec3f* lineBPointA, Vec3f* lineBPointB,
```